### PR TITLE
Fn render mutator path index tests

### PR DESCRIPTION
**Pull request by natasha41575**
Some of #1558, kpt fn render e2e tests

Mutator function generates new resources with `path` and `index` annotation specified. New resource declared in
- Current package
- Ancestor pacakge
- Descendent package
- Out of any package

### CODE REVIEW

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
++        - containerPort: 9876
+diff --git a/resources.yaml b/resources.yaml
+index 10cd18b..5b8e411 100644
+--- a/resources.yaml
++++ b/resources.yaml
+@@ -11,17 +11,29 @@
+ # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+-
```
Comment (Shell32-Natsu) [inline e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch:74]: We should avoid this kind of format changing in test.

File: e2e/testdata/fn-render/path-index-ancestor/db/Kptfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+apiVersion: kpt.dev/v1alpha2
+kind: Kptfile
+metadata:
+  name: db
+pipeline:
+  mutators:
+  - image: gcr.io/kpt-fn/starlark:unstable
```
Comment (Shell32-Natsu) [inline e2e/testdata/fn-render/path-index-ancestor/db/Kptfile:7]: Don't use unstable in kpt e2e test

Comment (natasha41575) [reply e2e/testdata/fn-render/path-index-ancestor/db/Kptfile:7]: What should I use instead? Unstable is the tag used in the already existing tests and I don't see another tag available in gcr.io/kpt-fn/starlark

Comment (droot) [reply e2e/testdata/fn-render/path-index-ancestor/db/Kptfile:7]: It is fine to use unstable for now. We will do a sweeping change in all the tests to use versioned images post m1 and I hope to have all the functions also tagged by then.

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch (hunk 1 above)
Comment (natasha41575) [reply e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch]: done

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+-#
+ # Licensed under the Apache License, Version 2.0 (the "License");
+ # you may not use this file except in compliance with the License.
+ # You may obtain a copy of the License at
+@@ -11,9 +9,19 @@
+ # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+-apiVersion: apps/v1
++apiVersion: apps/v1 # Copyright 2021 Google LLC
```
Comment (droot) [inline e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch:16]: This is strange. Something is broken some where. either in `fn render` or in one of the functions.
@Shell32-Natsu

File: e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  name: httpbin-gen
+source: |
+  httpbin_deployment = {
+     "apiVersion": "apps/v1",
+     "kind": "Deployment",
+     "metadata": {
+        "name": "httpbin",
+        "annotations": {
+          "config.kubernetes.io/index": 0,
+          "config.kubernetes.io/path": "../deployment_httpbin.yaml"
```
Comment (droot) [inline e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml:27]: Can we add a comment here.

File: e2e/testdata/fn-render/path-index-descendent/.expected/diff.patch [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+-# you may not use this file except in compliance with the License.
+ # You may obtain a copy of the License at
+ #
+ #      http://www.apache.org/licenses/LICENSE-2.0
+@@ -11,7 +7,7 @@
+ # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+-apiVersion: apps/v1
++apiVersion: apps/v1 # Copyright 2021 Google LLC
```
Comment (droot) [inline e2e/testdata/fn-render/path-index-descendent/.expected/diff.patch:62]: We need to figure out the issue with the license header.

File: e2e/testdata/fn-render/path-index-descendent/starlark-httpbin.yaml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  name: httpbin-gen
+source: |
+  httpbin_deployment = {
+     "apiVersion": "apps/v1",
+     "kind": "Deployment",
+     "metadata": {
+        "name": "httpbin",
+        "annotations": {
+          "config.kubernetes.io/index": 0,
+          "config.kubernetes.io/path": "./db/deployment_httpbin.yaml"
```
Comment (droot) [inline e2e/testdata/fn-render/path-index-descendent/starlark-httpbin.yaml:27]: add a comment.

File: e2e/testdata/fn-render/path-index-duplicate/.expected/config.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# One of the functions in the pipeline fails resulting in
+# non-zero exit code and no changes in the resources.
+exitCode: 1
+
+# This test case should fail, because if there is a duplicate path and index
+# with an existing resource, kpt should throw an error. Instead, kpt silently
+# fails to generate the resource and returns an exit code of 0.
```
Comment (droot) [inline e2e/testdata/fn-render/path-index-duplicate/.expected/config.yaml:21]: Can you pl. file a bug for this one, pointing to the test case. Also pl. update the spreadsheet.

**Review by droot — CHANGES_REQUESTED** [review]
Minor comments. Change is looking except the license headers issue (unrelated to this change).

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch (hunk 3 above)
Comment (natasha41575) [reply e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch:16]: Might be an issue with starlark or generators in general

See my comment here https://github.com/GoogleContainerTools/kpt/pull/1696#discussion_r608851162

File: e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml (hunk 4 above)
Comment (frankfarzan) [inline e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml:27]: @droot We need to fully specify the API here, it's part of the Functions Spec. Current docs:

https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/config-io.md

This is ambiguous:

`When writing Resources to a directory, the Resource SHOULD be written to the corresponding path relative to that directory.`

We need to define which `directory` this refers to.

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch [hunk 8]
```diff
@@ -0,0 +1,100 @@
+diff --git a/db/resources.yaml b/db/resources.yaml
+index af986cc..b0e8da9 100644
+--- a/db/resources.yaml
++++ b/db/resources.yaml
+@@ -1,20 +1,16 @@
+-# Copyright 2021 Google LLC
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch:6]: these diffs are a result of the functions removing comments (and therefore the licenses)

File: e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml (hunk 4 above)
Comment (droot) [reply e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml]: Agree. Filed an issue to track this https://github.com/GoogleContainerTools/kpt/issues/1718

**Review by droot — CHANGES_REQUESTED** [review]
Other than license/style issue, the PR is looking good.

I will try to see if we can debug the license/style issue today. If that takes longer, than we will have to address those in a follow up PR.

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch (hunk 3 above)
Comment (Shell32-Natsu) [reply e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch:18]: This should be an issue in kyaml.

https://github.com/GoogleContainerTools/kpt/issues/1706

We need to fix it there.

File: e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml (hunk 4 above)
Comment (frankfarzan) [reply e2e/testdata/fn-render/path-index-ancestor/db/starlark-httpbin.yaml]: How are we adding tests without knowing the expected behavior?

File: e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch (hunk 8 above)
[the remark this answers is not in the document]
Comment (Shell32-Natsu) [reply e2e/testdata/fn-render/path-index-ancestor/.expected/diff.patch:6]: I found that sometimes the starlark function will remove the licenses comments.

File: e2e/fn_test.go [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// +build docker
+
```
Comment (droot) [inline e2e/fn_test.go:1]: This is needed for guarding docker tests.

**Review by droot — CHANGES_REQUESTED** [review]
Minor comment.

**Review by droot — APPROVED** [review]

**Pull request merged**